test: add the possibility to run a test inside a network namespace

Tuetuopay opened the pull request:
For tests that do networking operations, this allows to have a clean-state network namespace and interfaces for each test. Mainly, this avoids "device or resource busy" errors when reusing the loopback interface across tests.

dave-tucker commented on test/integration-test/src/tests/smoke.rs:
```diff
@@ -4,8 +4,14 @@ use aya::{
+    Netns::new().exec(do_xdp)
+}
+
+fn do_xdp() {
```
```suggestion
    Netns::new().exec(|| {
```
You can use a closure to avoid creating a new function.

Tuetuopay replied:
Done in a956e8f

dave-tucker commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+
+use std::{
+    fs::File,
+    io::Write,
```
```suggestion
    io::Write as _,
```

Tuetuopay replied:
Done in 66c6ed7

dave-tucker commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+use std::{
+    fs::File,
+    io::Write,
+    os::fd::AsRawFd,
```
```suggestion
    os::fd::AsRawFd as _,
```

Tuetuopay replied:
Done in 66c6ed7

dave-tucker commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+    let mut stdin = bash.stdin.take().unwrap();
+    stdin.write_all(script.as_bytes()).unwrap();
+    stdin.write_all(b"\n").unwrap();
+    drop(stdin);
```
Why the explicit drop of stdin?

Tuetuopay replied:
Otherwise stdin does not get closed, and the bash we execute does not close terminate due to a missing EOF, deadlocking in the process.

Tuetuopay replied:
Resolving conversation as this code has been dropped.

dave-tucker commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
+        let name = format!("aya-test-{}", COUNTER.fetch_add(1, Ordering::Relaxed));
+        script(&format!(
+            "
+            ip netns add {name}
+            ip -netns {name} link set lo up
+        "
```
Formatting here looks off

Tuetuopay replied:
sadly rustfmt won't let me do the following. fixed in 199d5ca

```rust
        script(&format!("
            ip netns add {name}
            ip -netns {name} link set lo up
        "));
```

dave-tucker commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,76 @@
+
+impl Netns {
+    pub fn new() -> Self {
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
```
Would it make sense to move `COUNTER` outside of Netns?
Wondering if we ever run in parallel whether 2 tests could pick netns 0 at the same time.

Tuetuopay replied:
I don't follow you -- this is a static atomic, so as long as it's the same process, no two test will pick the same number. And since those integration tests are unit tests in the cargo sense, that's what happens.

Moving it outside of netns (out of `new` for that matter) would only make it accessible to foreign code (respecting visibility ofc)

tamird replied:
we might want to include the pid of the current process in the netns name to avoid concurrent processes from picking the same one.

dave-tucker reviewed (changes requested):
Looking good! A couple of nits, but excited to get this merged.

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,74 @@
+
+impl Drop for Netns {
+    fn drop(&mut self) {
+        script(&format!("ip netns del {}", self.name))
```
should the signature of script be

```
fn script<T: AsRef<[u8]>, I: IntoIterator<Item=T>>(script: I)
```
?

Tuetuopay replied:
Unless we want to pass binary data to the script (which I highly doubt), I'd prefer `T: AsRef<str>` instead to ensure _at least_ utf8. As for why a plain string and not `script: impl IntoIterator<Item=T>` where `T: AsRef<str>`, it all comes down to convenience. (see your further comment).

Tuetuopay replied:
Resolving as the script function has been dropped.

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,74 @@
+fn script(script: &str) {
+    // We pipe stdout/err to capture it and print it ourselves, so that it gets captured by the
+    // test harness
+    let mut bash = Command::new("/bin/bash")
```
why do we need to shell out to bash rather than use `Command::new("ip", ...)` multiple times?

Tuetuopay replied:
Pure convenience, and let's be honest, laziness.

This allows to write setup scripts without repeating the `Command::new` boilerplate code every time. Since this is made to do some setup, it allows to make readable setup scripts. Here only to create the network namespace, but it could be used by a test to create veths, install IPs, etc. And having 6 invocations of `Command::new(..).args(..).output().unwrap();` is just unwieldy.

This comes out of what we do at work (granted for actual applications and not libraries), where having an easy way to write test setup goes a long way for people to actually write useful tests.

tamird replied:
This laziness comes at a cost -- for example your current netns usage that executes two commands (one to create the network ns and another to `up` it) can fail to run the second half, and the Drop impl won't run in that case. I think we want to be more scrupulous here, not less.

Tuetuopay replied:
Indeed, good point. Though I still stand that for tests themselves this helper for running a script is useful.

Properly split with cleanups in f4c3968. It does so by creating the handle asap so that drop gets called if link up fails.

tamird replied:
I buy that we want to make this convenient for callers, but I don't think we should do that by using bash. Do we really want to end up with word splitting issues in our tests because we decided bash was an appropriate language?

Better would be to make this function's signature:
```rust
fn script<T: AsRef<OsStr>, I: IntoIterator<Item=T>>(script: impl IntoIterator<Item = I>)
```
note that this is different from my earlier suggestion in two ways:
- [u8] -> OsStr to match the bounds on various methods on `Command`
- `script` is now an iterator of iterators, so that this can be called as:
```rust
script([
  &["echo", "hello world"], // Look ma, no word splitting!
  &["rm", "-rf", "/"]
])
```

stepping back even further: we really really want to discourage shelling out in our tests because we want to drive toward having virtualized tests that exercise different kernel versions, and we don't want the userland of those VMs to become load bearing. We want to build the integration tests once, and then run them on a bunch of different VMs which ideally just have the kernel and little else.

Could we do this NetNS stuff using [the netns_rs crate](https://docs.rs/netns-rs/latest/netns_rs/) which doesn't depend on any userland at all?

Tuetuopay replied:
As requested:

- dropped the script function (well, not requested per se, but strongly implied)
- dropped the usage of iproute2, both for netns creation and for link up

Tuetuopay commented:
> we might want to include the pid of the current process in the netns name to avoid concurrent processes from picking the same one.

great idea. done in 3a44804

Tuetuopay commented:
@tamird I'll remove that script function. Something more ad-hoc will be added in the future when needed (even in my xdp maps pr I don't need it).

Good idea for the netns crate. Especially since it handles the rollback to the original netns, it'll remove the thread we spawn.

As for userspace, I think that `iproute2` is ubiquitous enough that it's not a big constraint to require it to run tests. Especially since we use it to up a link, that even `busybox`'s implementation of the `ip` command can handle. Furthermore, I really don't want to bring the complexity of manual netlink in the codebase for something as simple as upping `lo`. I'll have a look at the raw netlink code aya already has, and see if it can be added without too much code.

tamird commented on Cargo.toml:
```diff
@@ -71,6 +71,7 @@ integration-ebpf = { path = "test/integration-ebpf", default-features = false }
 lazy_static = { version = "1", default-features = false }
 libc = { version = "0.2.105", default-features = false }
 log = { version = "0.4", default-features = false }
+netns-rs = "0.1"
```
could you do the usual { version = .., default-features = false } please?

Tuetuopay replied:
Sure. Done in 8ec9f73

Tuetuopay replied:
Sure. Done in 8ec9f73

tamird commented on test/integration-test/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ aya-log = { workspace = true }
 aya-obj = { workspace = true }
 libc = { workspace = true }
 log = { workspace = true }
+netns-rs.workspace = true
```
please use the prevailing style

Tuetuopay replied:
Sure. Done in 8ec9f73

tamird commented on aya/src/sys/netlink.rs:
```diff
@@ -240,6 +240,32 @@ pub(crate) unsafe fn netlink_find_filter_with_name(
+    req.if_info.ifi_change = IFF_UP as u32;
+
+    sock.send(&bytes_of(&req)[..req.header.nlmsg_len as usize])?;
+    sock.recv()?;
```
do we need to check what we got back?

Tuetuopay replied:
AFAIK no. For an up request, we get back a simple ACK message with an error message, containing an error value of zero if successful, all of which is already checked in `NetlinkSocket::recv`.

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+use netns_rs::{get_from_current_thread, NetNs};
+
+pub struct Netns {
+    pub name: String,
```
do we need everything to be pub? seems like maybe none of it needs to be?

Tuetuopay replied:
Indeed it does not. Habit of logging everything. Done in 8ec9f73

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+use libc::if_nametoindex;
+use netns_rs::{get_from_current_thread, NetNs};
+
+pub struct Netns {
```
let's call this `NetNsGuard` to signal that it has important drop semantics?

Tuetuopay replied:
Sure. Done in 8ec9f73

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+
+        // Create and enter netns
+        let ns = NetNs::new(&name).unwrap_or_else(|e| panic!("Failed to create netns {name}: {e}"));
+        ns.enter().expect("Failed to enter network namespace");
```
nit: include the name in all the messages?

Tuetuopay replied:
Yup, this one slipped through. Done in 8ec9f73

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+        let ns = NetNs::new(&name).unwrap_or_else(|e| panic!("Failed to create netns {name}: {e}"));
+        ns.enter().expect("Failed to enter network namespace");
+        let netns = Self {
+            old_ns: get_from_current_thread().expect("Failed to get current netns"),
```
if this fails we leak the created namespace; consider doing this before creating the new one

Tuetuopay replied:
Done in 8ec9f73

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+        unsafe {
+            let idx = if_nametoindex(lo.as_ptr());
+            if idx == 0 {
+                panic!("Interface `lo` not found in netns {}", netns.name);
```
> On success, if_nametoindex() returns the index number of the
       network interface; on error, 0 is returned and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to
       indicate the error.

please emit the value of [`std::io::Error::last_os_error`](https://doc.rust-lang.org/stable/std/io/struct.Error.html#method.last_os_error)

Tuetuopay replied:
Done in 8ec9f73

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,61 @@
+            if idx == 0 {
+                panic!("Interface `lo` not found in netns {}", netns.name);
+            }
+            netlink_set_link_up(idx as RawFd)
```
RawFd is wrong here

Tuetuopay replied:
Why so? I copied this `as RawFd` from other uses of `if_nametoindex` in aya: see [`Xdp::attach`](https://github.com/aya-rs/aya/blob/main/aya/src/programs/xdp.rs#L99), that calls [`Xdp::attach_to_ifindex`](https://github.com/aya-rs/aya/blob/main/aya/src/programs/xdp.rs#L120) (uses `as RawFd`) which then calls [`aya::sys::netlink_set_xdp_fd`](https://github.com/aya-rs/aya/blob/main/aya/src/sys/netlink.rs#L26).

So if `RawFd` is wrong here, it's also wrong in the place I mentioned.

tamird replied:
I don't know about those other cites, but I think we can agree that an interface index is not a file descriptor?

Tuetuopay replied:
> I don't know about those other cites, but I think we can agree that an interface index is not a file descriptor?

Huh the sun hit hard, I didn't even realize this. Indeed that other use is wrong. Changing to `as i32`.

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -10,23 +11,26 @@ use aya::netlink_set_link_up;
         static COUNTER: AtomicU64 = AtomicU64::new(0);
         let pid = std::process::id();
         let name = format!("aya-test-{pid}-{}", COUNTER.fetch_add(1, Ordering::Relaxed));
+        let old_ns = get_from_current_thread().expect("Failed to get current netns");
```
sorry to harp on this more: make this the first thing in the function?the preceding two lines are infallible.

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -10,23 +11,26 @@ use aya::netlink_set_link_up;
+impl NetNsGuard {
     pub fn new() -> Self {
         static COUNTER: AtomicU64 = AtomicU64::new(0);
         let pid = std::process::id();
```
`use std::process` for consistency

Tuetuopay replied:
Done

Tuetuopay commented:
@tamird last comments taken into account. I squashed all commits in a single one (it ends up pretty short, to not pollute the log with useless commits) and rebased on `main`.

Thanks for the swift reviews!

tamird commented on test/integration-test/src/utils.rs:
```diff
@@ -0,0 +1,70 @@
+
+        // Create and enter netns
+        let ns = NetNs::new(&name).unwrap_or_else(|e| panic!("Failed to create netns {name}: {e}"));
+        ns.enter()
```
We need to create the NetNsGuard before this to ensure cleanup.

Tuetuopay replied:
Indeed. Fixed.

tamird commented:
@alessandrod PTAL since this technically affects public API.

tamird commented:
Per discord discussion we don't need API review here.
